java-frontend/src/test/java/org/sonar/java/metrics/MetricsComputerTest.java

```diff
@@ -0,0 +1,99 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class MetricsComputerTest implements MetricsScannerContext {
+
+  MetricsComputer mc = new MetricsComputer();
+
+  @Override
+  public MetricsComputer getMetricsComputer() {
+    return mc;
+  }
+
+  @SuppressWarnings({"unused", "unchecked"})
+  @Test
+  void testMetricsPresence() throws NoSuchFieldException, SecurityException, IllegalArgumentException, IllegalAccessException {
+    CompilationUnitTree cut = JParserTestUtils.parse(
+      "class A {" +
+        "  Object foo(){" +
+        "    if(a) { " +
+        "      for(int i=0; i<3; i++) { if(i==2){ return null; } }" +
+        "      for(int i : new int[]{1,2}) { if(i==2){ return null; } }" +
+        "      return new Object();   " +
+        "    };" +
+        "  } " +
+        "}");
+    MethodTree methodTree = (MethodTree) ((ClassTree) cut.types().get(0)).members().get(0);
+    mc.getComplexityNodes(methodTree).size();
+    var complexityNodes = (Map<Integer, ?>) getValue("methodComplexityNodes");
+    assertThat(complexityNodes).containsKey(methodTree.hashCode());
+    
+    mc.methodComplexity(methodTree);
+    var methodComplexity = (Map<Integer, ?>) getValue("methodComplexity");
+    assertThat(methodComplexity).containsKey(methodTree.hashCode());
+    
+    mc.getNumberOfAccessedVariables(methodTree);
+    var methodNumberOfAccessedVariables = (Map<Integer, ?>) getValue("methodNumberOfAccessedVariables");
+    assertThat(methodNumberOfAccessedVariables).containsKey(methodTree.hashCode());
+    
+    mc.linesOfCode(methodTree);
+    var treeLinesOfCode = (Map<Integer, ?>) getValue("treeLinesOfCode");
+    assertThat(treeLinesOfCode).containsKey(methodTree.hashCode());
+    
+    mc.numberOfStatements(methodTree);
+    var treeNumberOfStatements = (Map<Integer, ?>) getValue("treeNumberOfStatements");
+    assertThat(treeNumberOfStatements).containsKey(methodTree.hashCode());
+    
+    mc.numberOfCommentedLines(cut);
+    var treeNumberOfCommentedLines = (Map<Integer, ?>) getValue("treeNumberOfCommentedLines");
+    assertThat(treeNumberOfCommentedLines).containsKey(cut.hashCode());
+    
+    mc.noSonarLines(cut);
+    var treeNoSonarLines = (Map<Integer, ?>) getValue("treeNoSonarLines");
+    assertThat(treeNoSonarLines).containsKey(cut.hashCode());
+    
+    mc.compilationUnitComplexity(cut);
+    var compilationUnityComplexity = (Map<Integer, ?>) getValue("compilationUnityComplexity");
+    assertThat(compilationUnityComplexity).containsKey(cut.hashCode());
+    
+    mc.methodNestingLevel(methodTree);
+    var methodNestingLevel = (Map<Integer, ?>) getValue("methodNestingLevel");
+    assertThat(methodNestingLevel).containsKey(methodTree.hashCode());
+    
+  }
+
+  Object getValue(String fname) throws NoSuchFieldException, SecurityException, IllegalArgumentException, IllegalAccessException {
+    Field field = mc.getClass().getDeclaredField(fname);
+    field.setAccessible(true);
+    return field.get(mc);
+  }
```

Comment:
Maybe we would be better off by using a package-visible getter annotated with VisibleForTesting rather than playing with reflection